_posts/2023-07-15-gams-blog.md

```diff
@@ -0,0 +1,458 @@
+  0.813018 seconds (32.93 M allocations: 1.476 GiB, 38.89% gc time)
+
+````
+
+With a little effort, we can realize that the for loops are equivalent to
+treating each of the `IJK`, `JKL`, and `KLM` lists as a table in a database
+and performing an inner join across the similar indices.
+
+The blog post hints at this, saying "The reason for GAMS superior performance
+in this example is the use of relational algebra." But relational algebra,
+while not built-in to JuMP, is not unique to the GAMS modeling language. In
+Julia, we can use the `DataFrames` library.
+
+## The _DataFrames_ formulation
+
+The first step is to load the data as a dataframe instead of a list of tuples:
+
+````julia
+import DataFrames
+function parsefile_dataframe(filename::String, indices)
+    list = parsefile_tuple(filename)
+    return DataFrames.DataFrame(
+        [index => getindex.(list, i) for (i, index) in enumerate(indices)]...
+    )
+end
+struct DataFrameData
+    I::Vector{String}
+    IJK::DataFrames.DataFrame
+    JKL::DataFrames.DataFrame
+    KLM::DataFrames.DataFrame
+    function DataFrameData(n::Int)
+        return new(
+            ["i$i" for i in 1:n],
+            parsefile_dataframe(joinpath(DATA_DIR, "data_IJK_$n.json"), (:i, :j, :k,)),
+            parsefile_dataframe(joinpath(DATA_DIR, "data_JKL.json"), (:j, :k, :l)),
+            parsefile_dataframe(joinpath(DATA_DIR, "data_KLM.json"), (:k, :l, :m)),
+        )
+    end
+end
+dataframe_data = DataFrameData(2_200);
+````
+
+Using the dataframe data structure, we can compactly write an equivalent JuMP
+formulation:
+
+````julia
+function dataframe_formulation(data::DataFrameData)
+    ijklm = DataFrames.innerjoin(
+        DataFrames.innerjoin(data.IJK, data.JKL; on = [:j, :k]),
+        data.KLM;
+        on = [:k, :l],
+    )
+    model = Model(Gurobi.Optimizer)
+    set_silent(model)
+    ijklm[!, :x] = @variable(model, x[1:size(ijklm, 1)] >= 0)
+    for df in DataFrames.groupby(ijklm, :i)
+        @constraint(model, sum(df.x) >= 0)
+    end
+    optimize!(model)
+    return model
+end
+````
+
+````
+dataframe_formulation (generic function with 1 method)
+````
+
+This formulation doesn't look like the nested summation mathematics that GAMS
+originally formulated their model as, but it arguably just as readable,
+particularly if the `IJKLM` columns were meaningfully related to the business
+logic.
+
+This is much faster, taking just over 0.1 seconds:
+
+````julia
+@time dataframe_formulation(dataframe_data);
+@time dataframe_formulation(dataframe_data);
+````
+
+````
+  0.151296 seconds (461.83 k allocations: 51.187 MiB, 15.93% gc time, 19.65% compilation time)
+  0.120108 seconds (435.75 k allocations: 49.386 MiB, 21.98% gc time)
+
+````
+
+## Scaling
+
+Let's now compare the four different formulations over a range of `n` values:
+
+````julia
+import Plots
+function timings()
+    N = [100, 200, 400, 700, 1_100, 1_600, 2_200, 2_900]
```

Comment:
2_900->2_800

Reply:
No, I think I really mean 2_900?

Reply:
You are correct! 2900 is right.
I skipped a number XD